File: editors/vscode/src/main/messages/Notifications.ts
import * as vscode from "vscode";
import { MIN_MAJOR, MIN_MINOR } from "../cli/Python";

const missingNotified = new Set<string>();
const unsupportedNotified = new Set<string>();

function workspaceKey(folder?: vscode.WorkspaceFolder): string {
  return folder?.uri.fsPath ?? "__global__";
}

function notifyOnce(
  key: string,
  notified: Set<string>,
  title: string,
  message: string,
): void {
  if (notified.has(key)) {
    return;
  }

  notified.add(key);
  void vscode.window.showWarningMessage(`${title}: ${message}`);
}

/** Show one-time python interpreter warnings per workspace. */
export class Notifications {
  static resetForTests(): void {
    missingNotified.clear();
    unsupportedNotified.clear();
  }

  static missingNotificationCount(): number {
    return missingNotified.size;
  }

  static unsupportedNotificationCount(): number {
    return unsupportedNotified.size;
  }

  static notifyMissingPython(folder?: vscode.WorkspaceFolder): void {
    notifyOnce(
      workspaceKey(folder),
      missingNotified,
      "Docsig requires a Python interpreter",
      `No Python ${MIN_MAJOR}.${MIN_MINOR}+ interpreter is configured for this project`,
    );
  }

  static notifyUnsupportedPython(folder?: vscode.WorkspaceFolder): void {
    notifyOnce(
      workspaceKey(folder),
      unsupportedNotified,
      "Docsig requires a newer Python interpreter",
      `The configured python version is below minimum of ${MIN_MAJOR}.${MIN_MINOR}`,
    );
  }
}

File: plugin/neovim/lua/docsig/notifications.lua
local python = require("docsig.python")

local M = {}

local missing_notified = {}
local unsupported_notified = {}

local function workspace_key(bufnr)
  local root = require("docsig.python").workspace_root(bufnr)
  return root or "__global__"
end

local function notify_once(key, notified, title, message)
  if notified[key] then return end

  notified[key] = true
  vim.notify(string.format("%s: %s", title, message), vim.log.levels.WARN)
end

function M.reset_for_tests()
  missing_notified = {}
  unsupported_notified = {}
end

function M.notify_missing_python(bufnr)
  notify_once(
    workspace_key(bufnr),
    missing_notified,
    "Docsig requires a Python interpreter",
    string.format(
      "No Python %d.%d+ interpreter is configured for this project",
      python.MIN_MAJOR,
      python.MIN_MINOR
    )
  )
end

function M.notify_unsupported_python(bufnr)
  notify_once(
    workspace_key(bufnr),
    unsupported_notified,
    "Docsig requires a newer Python interpreter",
    string.format(
      "The configured python version is below minimum of %d.%d",
      python.MIN_MAJOR,
      python.MIN_MINOR
    )
  )
end

return M
